File: backend/BeElite/src/main/java/com/salesianos/triana/BeElite/dto/Program/ProgramDto.java
package com.salesianos.triana.BeElite.dto.Program;

import com.salesianos.triana.BeElite.model.Program;
import jakarta.transaction.Transactional;
import lombok.Builder;

import java.time.LocalDate;

@Builder
public record ProgramDto(
        String coachName,
        String coachUsername,
        String programName,
        String programDescription,
        String programPicUrl,
        LocalDate createdAt,
        int numberOfSessions,
        int numberOfAthletes,
        boolean isVisible) {

    @Transactional
    public static ProgramDto of(Program p) {
        return ProgramDto.builder()
                .coachName(p.getCoach().getName())
                .coachUsername(p.getCoach().getUsername())
                .programName(p.getProgramName())
                .programDescription(p.getDescription())
                .programPicUrl(p.getImage())
                .createdAt(p.getCreatedAt())
                .numberOfSessions(p.getWeeks() != null ?
                        p.getWeeks().stream().flatMap(week -> week.getSessions().stream()).mapToInt(session -> 1).sum()
                        : 0)
                .numberOfAthletes(p.getAthletes() != null ?
                        p.getAthletes().size()
                        : 0)
                .isVisible(p.isVisible())
                .build();
    }

    public static ProgramDto empty() {
        return ProgramDto.builder().build();
    }

}

File: frontend/be_elite/lib/models/Program/program_dto.dart
class ProgramDto {
	String? coachName;
	String? coachUsername;
	String? programName;
	String? programDescription;
	String? programPicUrl;
	String? createdAt;
	int? numberOfSessions;
	int? numberOfAthletes;
	bool? isVisible;

	ProgramDto({
		this.coachName, 
		this.coachUsername, 
		this.programName, 
		this.programDescription, 
		this.programPicUrl, 
		this.createdAt, 
		this.numberOfSessions, 
		this.numberOfAthletes, 
		this.isVisible, 
	});

	factory ProgramDto.fromJson(Map<String, dynamic> json) => ProgramDto(
				coachName: json['coachName'] as String?,
				coachUsername: json['coachUsername'] as String?,
				programName: json['programName'] as String?,
				programDescription: json['programDescription'] as String?,
				programPicUrl: json['programPicUrl'] as String?,
				createdAt: json['createdAt'] as String?,
				numberOfSessions: json['numberOfSessions'] as int?,
				numberOfAthletes: json['numberOfAthletes'] as int?,
				isVisible: json['isVisible'] as bool?,
			);

	Map<String, dynamic> toJson() => {
				'coachName': coachName,
				'coachUsername': coachUsername,
				'programName': programName,
				'programDescription': programDescription,
				'programPicUrl': programPicUrl,
				'createdAt': createdAt,
				'numberOfSessions': numberOfSessions,
				'numberOfAthletes': numberOfAthletes,
				'isVisible': isVisible,
			};
}
